src/modules/navigate.ts

```diff
@@ -232,3 +229,39 @@ export async function performNavigation(
 		window.history.back();
 	}
 }
+
+/**
+ * Undo the most recent visit: revert history entry and remove animation classes.
+ * @param visit The visit to abort.
+ * @returns void
+ * @internal
+ */
+export function undo(this: Swup, visit: Visit): void {
+	const state = (window.history.state as HistoryState) || {};
+
+	// Only undo currently running visits
+	if (!this.navigating || !visit || visit.done) return;
+
+	// Only undo visits that can be aborted, i.e. haven't started entering yet
+	if (!visit.abortable) return;
+
+	// Aborting most recent visit? Undo history and url bar changes
+	if (state.visit === visit.id) {
+		this.currentPageUrl = visit.from.url + visit.from.hash;
+		this.navigating = false;
+		this.hooks.callSync('visit:undo', visit, undefined, () => {
+			// Remove animation classes
+			this.classes.clear();
+			// Undo history and url bar changes
+			if (state.action === 'replace') {
+				updateHistoryRecord(visit.from.url + visit.from.hash);
+			} else {
+				window.history.back();
+			}
+		});
+	}
+
+	// Only set this here so that hooks still get called
+	// Probably set in visit.abort() right after visit.undo(), but let's make sure
```

Comment:
Needs adjustment to updated architecture. `visit.abort(true)` instead of `visit.undo()`